main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala

```diff
@@ -1201,87 +1199,103 @@ object GenExpression {
+        "()Ljava/util/function/Function;", false)
+      // stack: [index, acc, tupleElement*, terms, index, function]
+      // call the function with a new object as argument
+      visitor.visitTypeInsn(NEW, JvmType.Object);
+      // stack: [index, acc, tupleElement*, terms, index, function, object]
+      visitor.visitInsn(DUP)
+      // stack: [index, acc, tupleElement*, terms, index, function, object, object]
+      visitor.visitMethodInsn(INVOKESPECIAL, JvmType.Object.name.toInternalName, "<init>", AsmOps.getMethodDescriptor(List(), JvmType.Void))
+      // stack: [index, acc, tupleElement*, terms, index, function, object]
+      // prepare call to apply
+      visitor.visitInsn(SWAP)
+      // stack: [index, acc, tupleElement*, terms, index, object, function]
+      visitor.visitMethodInsn(INVOKEVIRTUAL, JvmName.Function.toInternalName, "apply",
+        "(Ljava/lang/Object;)Lflix/runtime/ProxyObject;", false)
+      // stack: [index, acc, tupleType, terms, index, tupleElement]
+
+      // prepare for recursion
+      visitor.visitInsn(DUP_X2)
+      // stack: [index, acc, tupleElement*, terms, index, tupleElement]
+      visitor.visitInsn(POP)
+      // stack: [index, acc, tupleElement*, terms, index]
+      // increase index by one
+      visitor.visitInsn(ICONST_1)
+      // stack: [index, acc, tupleElement*, terms, index, 1]
+      visitor.visitInsn(IADD)
+      // stack: [index, acc, tupleElement*, terms, index+1]
+      // recursion
+      visitor.visitJumpInsn(GOTO, tupleLoop)
 
-        visitLabel(exitTupleLoop)
-      } */
+      visitor.visitLabel(exitTupleLoop)
+      // stack: [index, acc, tupleElement*, terms, index]
+
+
+      // After exiting the loop, we have all the elements to construct the current tuple
+      visitor.visitInsn(POP2)
+      // stack: [index, acc, tupleElement*]
+
+      // Extract the type of elements that are in the tuple
+      val tupleElmsTypes = root.relations(pred).attr.map(_.tpe)
+      // Create a new tuple object
+      val classType = JvmOps.getTupleClassType(MonoType.Tuple(tupleElmsTypes))
+      visitor.visitTypeInsn(NEW, classType.name.toInternalName)
+      // stack: [index, acc, tupleElement*, tupleType]
+      val constructorDescriptor = AsmOps.getMethodDescriptor(??? /* TODO: this type should be the first type accepted by function f */, JvmType.Void)
+      visitor.visitMethodInsn(INVOKESPECIAL, classType.name.toInternalName, "<init>", constructorDescriptor, false)
+      // stack: [index, acc] // TODO: at this point we lost the tuple... it should be created before the loop, and kept at the second position on the stack when recursing
 
       // Now we have the tuple on the top of the stack
       // stack: [index, acc, tuple]
 
       // Getting the folded function
-      readVar(f, ???, visitor)
+      readVar(f, fType, visitor)
       // stack: [index, acc, tuple, f]
 
       // Call the function
-      ??? // TODO
+      ??? // TODO: how to do this cleanly?
```

Comment:
Maybe instead of having vars directly on FixpointSolve, you could let it have Expression.Var?

Reply:
That should indeed definitely be Expression.Var rather than a symbol + a type

Reply:
I am sorry that bytecode generation is such a nightmare 🤢